library/data/list/array_list.lean

```diff
@@ -0,0 +1,163 @@
+import init.data.quot
+universe u
+
+def defined_array.equiv (α) {n} (s : set (fin n)) (a b : array α n) := ∀ x ∈ s, a^.read x = b^.read x
+
+def defined_array.setoid (α) {n} (s : set (fin n)) : setoid (array α n) :=
+⟨defined_array.equiv α s,
+ λa x _, rfl,
+ λa b ab x h, eq.symm (ab x h),
+ λa b c ab bc x h, eq.trans (ab x h) (bc x h)⟩
+
+def defined_array (α) {n} (s : set (fin n)) := quotient (defined_array.setoid α s)
+
+def defined_array.lift_on {α n s} (q : @defined_array α n s) (β) (f : array α n → β)
+  (h : ∀ (a b : array α n), (∀ x ∈ s, a^.read x = b^.read x) → f a = f b) : β :=
+@quotient.lift _ β (defined_array.setoid α s) f h q
+
+def defined_array.read {α n s} (q : @defined_array α n s) (x) (h : x ∈ s) : α :=
+q^.lift_on (Π x ∈ s, α) (λa x _, a^.read x) (λ a b ab, funext $ λ x, funext $ λh, ab x h) x h
+
+def defined_array.write {α n s} (q : @defined_array α n s) (x) (h : x ∈ s) (v : α) : defined_array α s :=
+q^.lift_on _ (λa, @quotient.mk _ (defined_array.setoid α s) $ a^.write x v)
+  (λa b h, quot.sound (λi hi, show (if x = i then v else a^.read i) = (if x = i then v else b^.read i), by rw h i hi))
+
+def array_list.valid (size capacity) : set (fin capacity) := λ ⟨x, h⟩, x < size
+
+structure array_list (α : Type u) :=
+(capacity length : nat)
+(data : defined_array α (array_list.valid length capacity))
+(len_le : length ≤ capacity)
+
+def array_list.mk_data {α length capacity} (a : array α capacity) : defined_array α (array_list.valid length capacity) :=
+@quotient.mk _ (defined_array.setoid α (array_list.valid length capacity)) a
+
+local notation `⟦`:max a `⟧` := array_list.mk_data a
+
+def mk_array_list {α} [inhabited α] (capacity := 10) : array_list α :=
+{ capacity := capacity,
+  length := 0,
+  data := ⟦mk_array capacity (default α)⟧,
+  len_le := nat.zero_le capacity }
+
+def array.to_array_list {α n} (l : array α n) : array_list α :=
+{ capacity := n,
+  length := n,
+  data := ⟦l⟧,
+  len_le := le_refl n }
+
+def mk_array_list_of_fn {α n cap} [inhabited α] (f : fin n → α) (H : n ≤ cap) : array_list α :=
+{ capacity := cap,
+  length := n,
+  data := ⟦{data := λ⟨x, _⟩, if h : x < n then f ⟨x, h⟩ else default α}⟧,
+  len_le := H }
+
+def array.to_array_list_cap {α n cap} [inhabited α] (arr : array α n) (H : n ≤ cap) : array_list α :=
+mk_array_list_of_fn arr^.read H
+
+namespace array_list
+
+def nil {α} : array_list α :=
+{ capacity := 0,
+  length := 0,
+  data := ⟦array.nil⟧,
+  len_le := le_refl 0 }
+
+lemma lt_capacity {α} (l : array_list α) {n} (h : n < l^.length) : n < l^.capacity :=
+nat.lt_of_lt_of_le h l^.len_le
+
+def read {α} (l : array_list α) (x : fin l^.length) : α :=
+let ⟨n, h⟩ := x in l^.data^.read ⟨n, l^.lt_capacity h⟩ h
+
+def to_array {α} (l : array_list α) : array α l^.length :=
+{data := l^.read}
+
+def write {α} (l : array_list α) : fin l^.length → α → array_list α :=
+λ ⟨n, h⟩ v, { l with data := l^.data^.write ⟨n, l^.lt_capacity h⟩ h v }
+
```

Comment:
> I was hoping that version 4 would work. But it doesn't achieve the destructive write - do you know why? 

The problem is that there are implicit references to `l` in the `data^.write` application.
They can be eliminated. I will try to fix this.

That being said, I also think `write4` is a great alternative for compiling `{s with ...}`. I will try to work on this too.